### PR TITLE
[ao][quant] adding gpu quant primitives to ao/experimental

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112242
* #112531

Summary: adding the primitives for gpu quantization to ao/experimental,
these api's have no BC guarantee's and are subject to change. Note I to
add the experimental folder to the inline dynamo/inductor file list to
get the tests to pass

Test Plan: python torch/ao/quantization/experimental/_gpu_quantization/test.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/ao/quantization/experimental/_gpu_quantization/quant_primitives.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+
+# taken from
+# https://github.com/mit-han-lab/smoothquant/blob/2f87951dacfb9238d8d657f52ae83a82a3c9ba0c/smoothquant/fake_quant.py#L26
+# and slightly modified
+def quantize_activation_per_token_absmax(t):
+    n_bits = 8
+    # if the shape of t is [B, N, K], the shape of scales will be [B, N, 1]
+
+    scales = t.abs().max(dim=-1, keepdim=True)[0]
```
Comment (vadimkantorov) [inline torch/ao/quantization/experimental/_gpu_quantization/quant_primitives.py:134]: should this instead use `amax` directly: `t.abs().amax(dim=-1, keepdim=True)`?

File: torch/ao/quantization/experimental/_gpu_quantization/test.py [hunk 2]
```diff
@@ -0,0 +1,1024 @@
+# mypy: ignore-errors
```
Comment (cpuhrsch) [inline torch/ao/quantization/experimental/_gpu_quantization/test.py:1]: Please move tests under the toplevel test folder

File: torch/ao/quantization/experimental/_gpu_quantization/utils.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+def get_model_size_in_bytes(model):
+    s = 0
+    for p in model.parameters():
+        s += p.nelement() * p.element_size()
+    for b in model.buffers():
+        s += b.nelement() * b.element_size()
+    return s
+
+
+class IntMMModule(torch.nn.Module):
```
Comment (cpuhrsch) [inline torch/ao/quantization/experimental/_gpu_quantization/utils.py:108]: Is this still needed?

File: torch/ao/quantization/experimental/_gpu_quantization/smoothquant.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+#             alpha=self.alpha)
+#         self.fold_weight()
+
+#     def set_debug_x_absmax(self):
+#         W_absmax = torch.max(torch.abs(self.weight.transpose(0, 1)), dim=1).values
+#         W_absmax = _max_reduce(W_absmax)
+#         self.x_running_abs_max = W_absmax
+
+
+# # wrapper of https://github.com/facebookresearch/fairscale/
```
Comment (cpuhrsch) [inline torch/ao/quantization/experimental/_gpu_quantization/smoothquant.py:297]: There's a lot of commented-out code here - Should that be removed?

**Comment by cpuhrsch** [comment]
Test failures are real and mostly related to docs, API visibility and a dependency on matplotlib

File: torch/ao/quantization/experimental/_gpu_quantization/smoothquant.py [hunk 5]
```diff
@@ -0,0 +1,455 @@
+"""
+Testing out accuracy-only implementation of SmoothQuant
```
Comment (cpuhrsch) [inline torch/ao/quantization/experimental/_gpu_quantization/smoothquant.py:2]: As far as I know this isn't a fully faithful reimplement of smoothquant? Maybe we should revisit this later.

**Comment by cpuhrsch** [comment]
Marking this as WIP and converting it to draft

File: torch/ao/quantization/experimental/_gpu_quantization/utils.py (hunk 3 above)
Comment (HDCharles) [reply torch/ao/quantization/experimental/_gpu_quantization/utils.py:108]: no, i'll remove it

File: torch/ao/quantization/experimental/_gpu_quantization/test.py (hunk 2 above)
Comment (HDCharles) [reply torch/ao/quantization/experimental/_gpu_quantization/test.py:1]: OK

File: torch/ao/quantization/experimental/_gpu_quantization/quant_primitives.py (hunk 1 above)
Comment (HDCharles) [reply torch/ao/quantization/experimental/_gpu_quantization/quant_primitives.py:134]: yeah i can change that
